src/main/scala/com/wavesplatform/history/HistoryWriterImpl.scala

```diff
@@ -0,0 +1,82 @@
+package com.wavesplatform.history
+
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import com.google.common.util.concurrent.UncheckedExecutionException
+import scorex.account.Account
+import scorex.block.Block
+import scorex.block.Block.BlockId
+import scorex.transaction.History.BlockchainScore
+import scorex.transaction.ValidationError.CustomError
+import scorex.transaction.{History, HistoryWriter, ValidationError}
+import scorex.utils.ScorexLogging
+
+class HistoryWriterImpl(storage: HistoryStorage) extends History with HistoryWriter with ScorexLogging {
+
+  private val BlocksCacheSizeLimit: Int = 1000
+  private val blocksCache = CacheBuilder.newBuilder()
+    .maximumSize(BlocksCacheSizeLimit)
+    .build[Integer, Block](new CacheLoader[Integer, Block]() {
+    def load(height: Integer): Block = Block.parseBytes(storage.blockBodyByHeight.get(height)).get
+  }
+  )
+
+  override def appendBlock(block: Block): Either[ValidationError, Unit] = {
+    if ((height() == 0) || (this.lastBlock.uniqueId sameElements block.reference)) {
+      val h = height() + 1
+      storage.blockBodyByHeight.put(h, block.bytes)
+      storage.scoreByHeight.put(h, score() + block.blockScore)
+      storage.blockIdByHeight.put(h, block.uniqueId)
+      storage.heightByBlockId.put(block.uniqueId, h)
+      Right(())
+    } else {
+      Left(CustomError(s"Appending block ${block.json} which parent is not last block in blockchain"))
```

Comment:
With a full block of 100 transactions, this error message will be too long to be meaningful. I suggest mentioning just block id and maybe block parent's id.